src/main/java/edu/uwaterloo/ece658/service/DataTransmissionService.java

```diff
@@ -15,19 +27,124 @@
-//    }
-//
-//    public boolean updateS3key(String S3key, List<String> tags) {
-//        return false;
-//    }
-//
-//    public String download(String S3key) {
-//        return null;
-//    }
-//
-//    public boolean deleteImage(String S3key) {
-//        return false;
-//    }
+    @EJB
+    private UserFacade userFacade;
+
+    @EJB
+    private SubscriptionService subscriptionService;
+
+    @EJB
+    private PhotoFacade photoFacade;
+    
+    @EJB
+    private TagFacade tagFacade;
+
+    @EJB
+    private S3Util s3Util;
+
+    public URL getUploadURL(String s3Key) {
+        URL uploadURL = s3Util.getS3Url(s3Key, HttpMethod.PUT);
+        return uploadURL;
+    }
+    
+    public URL getDownloadURL(String S3key) {
+        URL downloadURL = s3Util.getS3Url(S3key, HttpMethod.GET);
+        return downloadURL;
+    }
+    
+    private Photo initializePhoto(
+            User user,
+            String S3key,
+            String Md5,
+            Integer photoSize, 
+            Date uploadDate, 
+            boolean isPrivatePhoto, 
+            List<Tag> tags)
+    {
+        Photo photo = new Photo();        
+        photo.setS3Key(S3key);
+        photo.setMd5(Md5);
+        photo.setSize(photoSize);
+        photo.setUploadedTime(uploadDate);
+        photo.setIsPrivate(isPrivatePhoto);
+        photo.setTags(tags);
+        photo.addUser(user);
+        photoFacade.create(photo);
+        return photo;
+    }
+    
+    private Tag createNewTag(String tagName) {
+        Tag tag = new Tag();
+        tag.setName(tagName);
+        tagFacade.create(tag);
+        return tag;
+    }
+    
+    private Tag initializeTag(String tagName) {
+        Tag tag = tagFacade.find(tagName);
+        if (tag == null) {
+            tag = createNewTag(tagName);
+        }
+        return tag;
+    }
+    
+    private List<Tag> initializeTags(List<String> tagNames) {
+        List<Tag> tags = new ArrayList<>();
+        
+        for (String tagName : tagNames) {
+            tags.add(initializeTag(tagName));
+        }
+        return tags;
+    }
+    
+    public boolean updateDatabaseWithPhotoInformation(
+            String username, 
+            String S3key,
+            String Md5,
+            Integer photoSize, 
+            Date uploadDate,
+            boolean isPrivatePhoto,
+            List<String> tagNames)
+    {
+        // store key in table with tags
+        List<Tag> tags = initializeTags(tagNames);
+        User user = userFacade.find(username);
+        Photo photo = initializePhoto(
+                        user,
+                        S3key,
+                        Md5,
+                        photoSize,
+                        uploadDate,
+                        isPrivatePhoto,
+                        tags);
+        
+        user.addUploadedPhoto(photo);
+        userFacade.edit(user);
+        
+        for (Tag tag : tags) {
+            tag.addPhoto(photo);
+            tagFacade.edit(tag);

```

Comment:
This edit() call is necessary, because tags might be newly created.

Reply:
They need to be updated because I just added a photo. But if that automatically updates the database then It can be removed